### PR TITLE
bugs

### PR DESCRIPTION
**Pull request by tomossomot**
[17](https://github.com/commerceblock/guardnode/issues/17): 
- check_for_bid_from_wallet include 0 conf txs
- check for `numTickets` fulfilled
- Fix client height not service height used

[20](https://github.com/commerceblock/guardnode/issues/20): 
- set `self.key`  to challenge signing key after picking up previously made bid

[18](https://github.com/commerceblock/guardnode/issues/18): 
- dont rescan on address key import

[16](https://github.com/commerceblock/guardnode/issues/16): 
- active request logic error fixed

other:
- feepubkey operations should use client ocean

### CODE REVIEW

**Comment by tomossomot** [comment]
commit `4a81f02` changes feepubkey address generation to be done by client chain  rather than service chain. This means that some users bids may be using feepubkeys generated by the wrong wallet. Something to keep in mind when debugging fee payments that are not received

**Comment by nkostoulas** [comment]
> commit `4a81f02` changes feepubkey address generation to be done by client chain rather than service chain. This means that some users bids may be using feepubkeys generated by the wrong wallet. Something to keep in mind when debugging fee payments that are not received.
> 
> As far as i can tell this is fine and will just require importing the corrcet key?

Ah how did we miss this. Still it's not a big problem yes. Only the public keys are recorded from the auction so the coordinator will pay out to the client address for this pub key. The users will have to find the pubkey they used, get the private key from this pubkey, convert it to client-chain private key and import that to the client chain wallet.

This is a bit fiddly unfortunately and the ocean wallet API does not help. It's a multi step process that requires (1) transforming the pubkey to a service chain address, (2) dumping the priv key for this address and (3) converting it from service to client chain priv key. (1)(3) have to be done outside the node using a script. I think @tomt1664 used a script like this in the past?

I used something like this in the past to convert keys from one chain to another:
```
import bitcoin as bc

keys = ['TiZcShQfZ9Uaix7a72HaejDJviBbuXugWxecDW9yQ4tELhJ6PcQa',]

for key in keys:
    print(key)
    num = bc.b58check_to_hex(key)
    print(bc.hex_to_b58check(num, 128))
    print()
```

File: guardnode/challenge.py [hunk 1]
```diff
@@ -138,79 +160,95 @@ def __init__(self, args):
...
-        
+
+        try:
+            # bid already made for current request
+            requestbids = self.service_ocean.getrequestbids(self.request["txid"])["bids"]
+            if any(bid["txid"] == self.bid_txid for bid in requestbids):
+                return False
+
+            # no tickets left
+            num_bids = len(self.service_ocean.getrequestbids(self.request["txid"])["bids"])
```
Comment (nkostoulas) [inline guardnode/challenge.py]: No need to do the rpc call twice.

File: guardnode/guardnode.py [hunk 2]
```diff
@@ -30,7 +30,7 @@ def parse_args():
     parser.add_argument('--nodelogfile', required=False, type=str, default=NODE_LOG_FILE_DEFAULT, help="Node log file destination")
 
     parser.add_argument('--bidpubkey', required=False, type=str, help="Guardnode winning bid public key")
-    parser.add_argument('--bidlimit', required=False, type=float, default="0.0", help="Guardnode upper bid limit")
+    parser.add_argument('--bidlimit', required=False, type=float, default="1000", help="Guardnode upper bid limit")
```
Comment (nkostoulas) [inline guardnode/guardnode.py]: Maybe we can decrease this further? We don't want people bidding if they are not aware of the limit.

File: guardnode/challenge.py [hunk 3]
```diff
@@ -138,79 +160,95 @@ def __init__(self, args):
...
                         if not self.await_challenge():
                             break   # request ended
-                        sleep(0.1) # seconds
+                        sleep(1) # seconds
                 else:
                     sleep(self.args.serviceblocktime)
             else:
-                self.logger.info("No active requests for genesis: {}".format(self.genesis))
+                self.no_request_msg_count += 1
+                if self.no_request_msg_count in expos: # gradually get more quiet
```
Comment (nkostoulas) [inline guardnode/challenge.py]: Interesting.

Comment (nkostoulas) [reply guardnode/challenge.py]: I guess this stops printing after 16*16 minutes? You could check if the root of the number is an integer if you want to extend further, or just leave it as is I guess.

File: guardnode/challenge.py (hunk 1 above)
Comment (tomossomot) [reply guardnode/challenge.py]: good spot, fixed

File: guardnode/guardnode.py (hunk 2 above)
Comment (tomossomot) [reply guardnode/guardnode.py]: hm okay, probably best to remain at 0 then.

File: guardnode/challenge.py (hunk 3 above)
Comment (tomossomot) [reply guardnode/challenge.py]: I like this a lot

**Comment by tomossomot** [comment]
Cheers for the layout for how to do the feepubkey wallet change. 

Added a test for the wallet change feature

Also noticed the 0 conf bid was incorrectly reading the feepubkey

**Pull request merged**